### PR TITLE
Chain 690 - initial support for depositing and withdrawing runes to contract

### PR DESCRIPTION
**Pull request by bflood366**
- includes support for depositing and withdrawals of runes
- there is a new runes receiver account that will receive all runes deposited to the exchange. This will hold the rune utxos and will be a signer on batch withdrawals in which runes are being spent for withdrawal purposes.
- Separately there is a separate token account to hold balances for each rune. The token id field for a rune holds the rune_id. That is how differentiate between a token account holding btc balances and one holding rune balances.
- the instructions for prepare and batch withdrawal have been updated. The changes include allowing a different fee account and fee address index if the fee for a withdrawal is in a different token than the token being withdrawn. 
- this contract won't work with the existing backend - once we change backend to work with this contract, part of deploying that backend to testnet would require us to deprecate the existing contract which would trigger updating the new contract - for now though I will keep this in a feature branch.
- this does not include any changes wrt the bonding curve user - that can be handled in a different PR

### CODE REVIEW

**Review by bholzman — APPROVED** [review]
Looks great!

**Review by bholzman — APPROVED** [review]
Looks good!

File: arch/contracts/exchange/model/src/state.rs [hunk 1]
```diff
@@ -399,6 +474,7 @@ pub fn validate_account(accounts: &[AccountInfo], index: u8, is_signer: bool, is
         return Err(ProgramError::Custom(ERROR_INVALID_ACCOUNT_INDEX));
     }
     let account = &accounts[index as usize];
+    msg!("validate account for {:?} {} {}", account_type, account.is_signer, account.is_writable);
```
[the remark this answers is not in the document]
Comment (bflood366) [reply arch/contracts/exchange/model/src/state.rs:477]: nope thx - removed

File: arch/contracts/exchange/model/src/instructions.rs [hunk 2]
```diff
@@ -12,6 +12,7 @@ pub enum ProgramInstruction {
     RollbackBatchSettlement(),
     RollbackBatchWithdraw(RollbackWithdrawBatchParams),
     SubmitBatchWithdraw(WithdrawBatchParams),
+    InitRuneReceiverState,
```
[the remark this answers is not in the document]
Comment (bflood366) [reply arch/contracts/exchange/model/src/instructions.rs:15]: ah yep, changed

File: arch/contracts/exchange/model/src/serialization.rs [hunk 3]
```diff
@@ -651,13 +691,16 @@ impl Codable for Event {
                     writer.write_u64(*balance)? +
                     writer.write_u32(*error_code)?
             }
-            Self::FailedWithdrawal { account_index, address_index, requested_amount, fee_amount, balance, error_code } => {
+            Self::FailedWithdrawal { account_index, address_index, fee_account_index, fee_address_index, requested_amount, fee_amount, balance, balance_in_fee_token, error_code, .. } => {
```
[the remark this answers is not in the document]
Comment (bflood366) [reply arch/contracts/exchange/model/src/serialization.rs:694]: odd I did not explicitly add the `..` there, I must of checked one of the IDE's fix suggestions at some point - good catch

File: arch/contracts/exchange/program/src/lib.rs [hunk 4]
```diff
@@ -163,48 +204,92 @@ pub fn submit_withdraw_batch(program_id: &Pubkey, accounts: &[AccountInfo], para
...
+            ).expect("cannot add change edict");
+        });
+
+        let runestone = Runestone {
+            edicts,
+            etching: None,
+            mint: None,
+            pointer: None,
+        };
+        msg!("runestone is {:?}", runestone);
```
[the remark this answers is not in the document]
Comment (bflood366) [reply arch/contracts/exchange/program/src/lib.rs:260]: removed

**Pull request merged**